File: apps/backendts/src/workflows/automated.ts
import { step } from "@restackio/ai/workflow";
import * as openaiFunctions from "@restackio/integrations-openai/functions";
import { openaiTaskQueue } from "@restackio/integrations-openai/taskQueue";
import * as functions from "../functions";
import zodToJsonSchema from "zod-to-json-schema";
import { todoSchema } from "../functions/todos/types";

export async function automatedWorkflow() {
  const hnData = await step<typeof functions>({}).toolHnSearch({
    query: "ai",
  });

  const todosJsonSchema = {
    name: "todos",
    schema: zodToJsonSchema(todoSchema),
  };

  const result = await step<typeof openaiFunctions>({
    taskQueue: openaiTaskQueue,
  }).openaiChatCompletionsBase({
    model: "gpt-4o-mini",
    userContent: `
      You are a personal assistant.
      Here is the latest hacker news data:
      ${JSON.stringify(hnData)}
      Create a todo for me to contact the founder with a one sentence summary of their product.`,
    jsonSchema: todosJsonSchema,
  });

  return result;
}

File: apps/backend-python/src/workflows/automated.py
from src.restack_ai.workflow import workflow, workflow_import
from datetime import timedelta
from pydantic import BaseModel
from src.functions.hn.schema import HnSearchInput

with workflow_import():
    from src.functions.openai.chat import openai_todos, FunctionInputParams
    from src.functions.hn.search import tool_hn_search
class WorkflowInputParams(BaseModel):
    pass

@workflow.defn(name="automatedWorkflow")
class AutomatedWorkflow:
    @workflow.run
    async def run(self, input: WorkflowInputParams):

        hn_results = await workflow.step(tool_hn_search, HnSearchInput(query="ai"), start_to_close_timeout=timedelta(seconds=10))

        user_content = f"You are a personal assistant. Here is the latest hacker news data: {str(hn_results)} Create a todo for me to contact the founder with a one sentence summary of their product"

        result = await workflow.step(openai_todos, FunctionInputParams(user_content=user_content), start_to_close_timeout=timedelta(seconds=10))

        return {"result": result}